teps/0008-trigger-crd.md

```diff
@@ -0,0 +1,434 @@
+which it is bound.
+
+### EventListener CRD
+EventListeners expose an addressable "Sink" to which incoming events in the form of HTTP requests
+are directed and process those requests with configuration from bound Triggers. EventListener has
+triggerSelector to select triggers based on labels and namespaceSelector to select namespaces where
+it searches for triggers.
+```
+apiVersion: v1alpha1
+kind: EventListener
+metadata:
+  name: my-el
+spec:
+  serviceAccountName: "blah"
+  namespaceSelector:
+     matchName: [ns-1, ns-2]
+     matchLabels:
+       - tekton-triggers: tekton-operator
+  triggerSelector:
+    matchLabels:
+      - eventlistener: tekton-operator
+```
+
+### Path based EventListener
+Inside EventListener, we will deduce Trigger to execute based on the path of the request URL.
+To refer to trigger resource ```foo``` in namespace ```bar```, we would have a url ```/bar?name=foo```.
+Instead of EventListener iterating through every Triggers that can be serve, EventListener
+will directly process the Trigger refer by URL. Similarly we can have url ```/bar?label=app-foo``` 
+where EventListener will only process all the Triggers matching labels  ```app-foo```.
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this TEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+We have two type of users:
+1. End users who requires his event to be processed.
+2. Operator/admin who managed EventListener.
+#### Trigger Ref inside the EventListener
+Instead of defining Trigger inside EventListener, Trigger resource will be created by end user.
+Trigger will contain serviceaccount, triggerbinding, triggertemplate and interceptor.ServiceAccount
+will be optional. If not defined, ServiceAccount of EventListener will be used.
+
+#### Specifying Namespaces where EventListener can serve
+Admin/Operator would give which namespaces EventListener can serve. In this way, EventListener can be used
+in Multitenant Scenarios.
+
+#### Path Base EventListener Implementation
+Path base EventListener will be implemented.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+<!--
+What are the risks of this proposal and how do we mitigate. Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the WGs or subproject.
+-->
+1. Security issue - Wide permission to EventListener. In this scenario EventListener
```

Comment:
can you describe a bit more detail what the wide permissions are? How can an admin manage it e.g. they may want to restrict write access to the Listener

Reply:
Yes some elaboration would be good here.  i.e. that a single event listener would either have to have permissions to each of the namespaces, or for convenience sake, get clusterrole permissions, but that can give the EL access to "system level namesapces", like where the apiserver live, if done naively.

Or that a user could purposefully configure a github webhook to get to a "restricted" namesapce instead of a namespace they had access to if they got a hold of the verification secret, etc.

Stuff along those lines.

Reply:
I rephrased and added how it will be mitigated. I was talking about EventListener having wide permissions,